### PR TITLE
Let us follow a path

### LINKED ISSUE #132: Allow User to Create Object Path
**Is your feature request related to a problem? Please explain.**
One thought is to enable a user to create an arbitrary path that an object could follow.

**Describe the solution you'd like**
Luxor and Cairo has many methods which support path creation. Using those would be the starting point to creating a proof of concept.

### PR DESCRIPTION
**Pull request by Wikunia**
## PR Checklist

If you are contributing to `Javis.jl`, please make sure you are able to check off each item on this list:

- [x] Did I update `CHANGELOG.md` with whatever changes/features I added with this PR?
- [ ] Did I update the `TODO.md` (if applicable)?
- [x] Did I make sure to only change the part of the file where I introduced a new change/feature?
- [ ] Did I cover all corner cases to be close to 100% test coverage (if applicable)?
- [x] Did I properly add Javis dependencies to the `Project.toml` + set an upper bound of the dependency (if applicable)?
- [x] Did I properly add test dependencies to the `test` directory (if applicable)?
- [ ] Did I check relevant tutorials that may be affected by changes in this PR?
- [x] Did I clearly articulate why this PR was made the way it was and how it was made?

**Link to relevant issue(s)**
Closes #132 

**How did you address these issues with this PR? What methods did you use?**
It is now possible with `SubAction(1:100, follow_path(star(O, 5)))` to make the parent action follow the outline of a star.
For this I've added a new field to the `SubAction` struct to compute the distances only once and added a `follow_path` function into the subaction animations. It borrows the ideas from the `Luxor.polyportion` function

follow_path

### CODE REVIEW

File: src/subaction_animations.jl [hunk 1]
```diff
@@ -289,3 +289,55 @@ function _sethue(video, action, subaction, rel_frame)
...
+"""
+function follow_path(points::Vector{Point}; closed = true)
+    (video, action, subaction, rel_frame) ->
+        _follow_path(video, action, subaction, rel_frame, points; closed = closed)
+end
+
+function _follow_path(video, action, subaction, rel_frame, points; closed = closed)
+    t = get_interpolation(subaction, rel_frame)
+    t -= floor(t)
+    if isapprox(t, 1.0, atol = 1e-4)
```
Comment (Wikunia) [inline src/subaction_animations.jl:318]: I should check whether `t` is an integer first as when `t == 1` I actually want it to be `1` not `0`.
Additionally I want to do the `t -= floor(t)` thing only if `closed` as this otherwise doesn't make sense and should probably throw an error

**Review by TheCedarPrince — CHANGES_REQUESTED** [review]
Some minor typos and clarification requests, but looks great! I had fun making this:

```julia
using Javis
using Animations
using Colors

function ground(args...)
        background("black")
        sethue("white")
end

  video = Video(800, 600)

    anim = Animation([0, 1], [0.0, 2.0], [sineio()])

    color_anim = Animation(
        [0, 0.5, 1], # must go from 0 to 1
        [Lab(colorant"red"), Lab(colorant"cyan"), Lab(colorant"red")],
        [sineio(), sineio()],
    )

    actions = [
        Action(
            frame_start:(frame_start + 149),
            (args...) -> star(O, 20, 5, 0.5, 0, :fill);
            subactions = [
                SubAction(1:150, anim, follow_path([Point(path_val, 200), Point(-path_val, -200)])),
                SubAction(1:150, color_anim, sethue()),
            ],
        ) for (frame_start, path_val) in zip(1:4:25, -300:100:300)
    ]

    javis(
        video,
        [BackgroundAction(1:180, ground), actions...],
        pathname = "star.gif"
    )
```

star

File: src/Javis.jl [hunk 2]
```diff
@@ -63,6 +63,8 @@ Create a video with a certain `width` and `height` in pixel.
 This also sets `CURRENT_VIDEO`.
 """
 function Video(width, height)
+    # some luxor functions need a drawing ;)
+    Drawing()
     video = Video(width, height, Dict{Symbol,Any}())
```
[the remark this answers is not in the document]
Comment (Wikunia) [reply src/Javis.jl:68]: cormullion is doing something similar to what we do with Action and Video such that we have `CURRENT_ACTION` as an example. All the drawing functions need a Drawing so if we call some drawing function before we call `javis` it will fail.
I think that this doesn't do any harm and makes it easier for the user.

File: src/subaction_animations.jl [hunk 3]
```diff
@@ -289,3 +289,63 @@ function _sethue(video, action, subaction, rel_frame)
...
+end
+
+function _follow_path(video, action, subaction, rel_frame, points; closed = closed)
+    t = get_interpolation(subaction, rel_frame)
+    # if not closed it should be always between 0 and 1
+    if !closed
+        t = clamp(t, 0.0, 1.0)
+    end
+    # if t is discrete and not 0.0 take the last point or first if closed
+    if rel_frame != 1 && isapprox_discrete(t)
```
[the remark this answers is not in the document]
Comment (Wikunia) [reply src/subaction_animations.jl:323]: the path you use is two points so `A` and `B` but the default kwarg is `closed = true` to make it easy to work with polygons basically. This means the movement is `A -> B -> A`. This basically checks: okay we are not in the first frame but we are discrete therefore we want to just return the end point instead of continuing with the computation of finding the point.

File: src/subaction_animations.jl [hunk 4]
```diff
@@ -289,3 +289,63 @@ function _sethue(video, action, subaction, rel_frame)
     color = get_interpolation(subaction, rel_frame)
     Luxor.sethue(color)
 end
+
+"""
+    follow_path(points::Vector{Point}; closed=true)
+
+Can be applied inside a subaction such that the object defined in the parent action follows a path.
+It takes a vector of point which can be created as an example by calling `circle(O, 50)` <- notice that the action is set to `:none` the default.
```
[the remark this answers is not in the document]
Comment (Wikunia) [reply src/subaction_animations.jl]: thx resolved

File: src/subaction_animations.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (TheCedarPrince) [reply src/subaction_animations.jl:324]: This makes sense! Thanks for explaining.

File: src/Javis.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (TheCedarPrince) [reply src/Javis.jl:72]: This also makes sense about why this would be required. Thanks for clarifying.

**Review by TheCedarPrince — APPROVED** [review]
Thanks for explaining more here. I approve and love this new feature. Excited to see how it can be used.
